src/python/pants/backend/docker/util_rules/docker_binary.py

```diff
@@ -149,14 +136,42 @@ async def find_docker(
             search_path=search_path,
         ),
     )
-    tools_path = ".shims"
-    extra_env = {"PATH": os.path.join(tools_path, tools.bin_directory)}
-    extra_input_digests = {tools_path: tools.digest}
+
+    # In addition to the docker-tool shims we generate above, we generate a shim for the
+    # `docker` CLI itself. This shim ensures that the _absolute path_ to the `.shim/bin`
+    # directory populated above is on the `PATH` for the CLI.
+    #
+    # It's important that we use the absolute path to `.shims/bin` because the use of a
+    # relative path raises an error. The error comes from the `LookPath` function in the
+    # `golang.org/x/sys/execabs` package, which the `docker` CLI started using as of
+    # https://github.com/docker/cli/commit/8d199d5bba9db46b6610bd959d815ce7197402b3. The
+    # error condition was added purposefully by the core Go team, and it's since been
+    # ported to `os/exec` in the Go stdlib (see https://github.com/golang/go/issues/43724).
+    #
+    # Using a shim script for the docker CLI here is ugly, but as far as I can tell it's
+    # the only way to dynamically invoke/inject the value of `$(pwd)` for the docker processes
+    # into an env value. Trying to set `{"PATH": "$(pwd)/.shims/bin"}` in the `env` field
+    # for the returned `DockerBinary` instance doesn't work because the values in that map
+    # are shell-escaped before they are written to `__run.sh`.
+    docker_shim_script = FileContent(
+        "__run_docker.sh",
+        textwrap.dedent(
+            f"""\
+            #!/bin/bash
+            export PATH="$(pwd)/{tools_path}/{tools.bin_directory}:${{PATH}}"
+            exec "{first_path.path}" "${{@}}"
+            """
+        ).encode("utf-8"),
+        is_executable=True,
+    )
+    docker_shim_digest = await Get(Digest, CreateDigest([docker_shim_script]))
+
+    shims_digest = await Get(Digest, MergeDigests([tools.digest, docker_shim_digest]))
+    extra_input_digests = {tools_path: shims_digest}
 
     return DockerBinary(
-        first_path.path,
+        f"{tools_path}/{docker_shim_script.path}",
         first_path.fingerprint,
```

Comment:
I feel like this value should probably change, but I'm not sure what it should be set to.